### ISSUE TITLE
Training failure when enabling cr-ctc-loss - possible grad_scale issue

### ISSUE THREAD

**Issue by redsheep913** [k2-fsa/icefall#2042]
@AmirHussein96 
Hi, I’m currently training the HENT-SRT model following the paper. I enabled RNN-T loss, CTC loss, and CR-CTC loss together as described. However, training fails due to what seems like gradient scaling (grad_scale) being too small, causing instability.  

Could you please advise on the recommended settings for grad_scale and loss weight hyperparameters when using cr-ctc-loss and ctc-loss simultaneously? Any tips to avoid training failures would be greatly appreciated!  

Thank you!

**Comment by AmirHussein96** [comment]
Hi @redsheep913 ,

Since you mentioned following the Hent-SRT training steps in [this guide](https://github.com/k2-fsa/icefall/blob/master/egs/multi_conv_zh_es_ta/ST/RESULTS.md), I assume you started with ASR pretraining — is that correct? Could you clarify whether this issue occurs during the ASR pretraining stage or later during ST training? Also please share is you are using same data as in the paper or some other data?

This error typically happens in CTC when the text sequence length is greater than or equal to the number of audio frames after downsampling. As a quick check, try disabling the CTC loss to see if the issue still persists. Also make sure you are using this filtering function (https://github.com/k2-fsa/icefall/blob/693f069de73fd91d7c2009571245d97221cc3a3f/egs/multi_conv_zh_es_ta/ST/zipformer_multijoiner_st/train.py#L1602). 
If this is indeed caused by CTC we then can dive deeper into the CTC reduction and try to get the exact samples that causes this issue so you can filter them out.

**Comment by danpovey** [comment]
The grad scale gets small when grads are large too many times in a row.  This can be due to many different forms of training instability.  The normal first thing to try when this happens is to decrease the learning rate.  Or if it happens while the LR is still warning up, you can modify the LR warm up schedule to make it slower.
Another thing to do is to run with --inf-check=True, assuming the recipe supports it, and look at which grads are getting large, or which "forward" quantities, in rarer cases, which can suggest what to change, e.g. changing initializations of different modules.  It will print out the names of all infinite things but what you want to pay attention to is the first one for each "bad" batch, which is generally possible to work out from the sequence of them if you think about how backprop works.  E.g. supposing layers.3.self_attn.in_proj grad is infinite, after that normally things from layers.3, layers,2, layers.1, layers.0, and the frontend will be printed out as well, after which you'll see some unrelated messages followed by, if it happens again, the next "bad" gradient.

**Comment by redsheep913** [comment]
Hi @AmirHussein96, @danpovey 
Yes, I have followed the guide, and the issue occurred during the first pretraining stage of the offline CR-CTC ASR.  
The model worked well with the RNN-T loss independently, so I attempted to enable both the CTC loss and CR-CTC loss.  
The `--inf-check=True` option was enabled during training, and the filtering function `remove_short_and_long_utt` was also confirmed to be enabled.
Since I am training with only one GPU (NVIDIA A10), I initially decreased the learning rate from 0.055 to 0.0005. I would appreciate any advice on whether this learning rate change is appropriate.

I am using data consisting of Taiwanese Hokkien audio, Taiwanese Hokkien transcripts, and parallel Mandarin Chinese transcripts.  
I changed the "en" label to "zh" to reflect Taiwanese Hokkien to Mandarin Chinese speech translation.  
For the tokenizer, I replaced the original BPE model with a byte-level BPE model, following the aishell bbpe training model [script](https://github.com/k2-fsa/icefall/blob/693f069de73fd91d7c2009571245d97221cc3a3f/egs/aishell/ASR/zipformer/train_bbpe.py#L4)

[hent_srt_asr_taiwanese_ctc_1114_train.log](https://github.com/user-attachments/files/23541092/hent_srt_asr_taiwanese_ctc_1114_train.log)

The training log with a base learning rate of 5e-04:
[hent_srt_asr_taiwanese_ctc_1114_train_inf_check.log](https://github.com/user-attachments/files/23541865/hent_srt_asr_taiwanese_ctc_1114_train_inf_check.log)

The training log with a base learning rate of 5e-05:
[log-train-lr-5e-05.log](https://github.com/user-attachments/files/23542430/log-train-lr-5e-05.log)

I have already adjusted the base learning rate from the original 0.055 down to 5e-04, and then further down to 5e-05. However, from the --inf-check logs, it seems that at the beginning of training there is consistently a warning message:
`WARNING [hooks.py:90] The sum of encoder_embed.out.bias.param_grad is not finite`.

Would you happen to have any ideas about what might be causing this issue? Any advice or suggestions would be very much appreciated.

Below is the output from `display_manifest_statistic` showing the cut information.



```
Cut statistics:

╒═══════════════════════════╤══════════╕

│ Cuts count: │ 139272 │

├───────────────────────────┼──────────┤

│ Total duration (hh:mm:ss) │ 76:16:41 │

├───────────────────────────┼──────────┤

│ mean │ 2.0 │

├───────────────────────────┼──────────┤

│ std │ 0.8 │

├───────────────────────────┼──────────┤

│ min │ 0.5 │

├───────────────────────────┼──────────┤

│ 25% │ 1.4 │

├───────────────────────────┼──────────┤

│ 50% │ 1.8 │

├───────────────────────────┼──────────┤

│ 75% │ 2.4 │

├───────────────────────────┼──────────┤

│ 99% │ 4.3 │

├───────────────────────────┼──────────┤

│ 99.5% │ 4.6 │

├───────────────────────────┼──────────┤

│ 99.9% │ 5.4 │

├───────────────────────────┼──────────┤

│ max │ 8.6 │

├───────────────────────────┼──────────┤

│ Recordings available: │ 139272 │

├───────────────────────────┼──────────┤

│ Features available: │ 139272 │

├───────────────────────────┼──────────┤

│ Supervisions available: │ 139272 │

╘═══════════════════════════╧══════════╛

SUPERVISION custom fields:

- translated_text (in 139272 cuts)

Speech duration statistics:

╒══════════════════════════════╤══════════╤══════════════════════╕

│ Total speech duration │ 76:16:41 │ 100.00% of recording │

├──────────────────────────────┼──────────┼──────────────────────┤

│ Total speaking time duration │ 76:16:41 │ 100.00% of recording │

├──────────────────────────────┼──────────┼──────────────────────┤

│ Total silence duration │ 00:00:00 │ 0.00% of recording │

╘══════════════════════════════╧══════════╧══════════════════════╛
```

**Comment by danpovey** [comment]
Firstly, that learning rate is way too small.  I'd normally decrease it only slightly, e.g. from 0.05 to 0.04, not by factors of 10 or more.  It's very sensitive to small differences in LR, with this optimizer.
Secondly: if it says "encoder_embed.out.bias.param_grad" is not finite only once or twice, it might be OK.  But I'd be surpised if that was always the first thing to be printed, as the grad would have to be already quite large near that point, and normally some of the later layers would also have non-finite grad, sometimes.  Anyway, this is a quite generic type of instability.
What to do to fix it depends on which batch this happens at and what the loss value looked like at the point when it diverged.

**Comment by danpovey** [comment]
The first thing I see in the log is:
```
2025-11-14 15:07:52,893 WARNING [hooks.py:90] The sum of ctc_output.1.weight.param_grad is not finite
2025-11-14 15:07:52,894 WARNING [hooks.py:90] The sum of ctc_output.1.bias.param_grad is not finite
2025-11-14 15:07:52,894 WARNING [hooks.py:77] The sum of ctc_output.1.grad[0] is not finite
2025-11-14 15:07:52,895 WARNING [hooks.py:77] The sum of ctc_output.0.grad[0] is not finite
```
.. this is not a normal error.  It's the kind of thing you might get if the number of symbols of output is more than the number of frames of CTC output.  These kinds of utterancers are normally filtered out somehow.

**Comment by csukuangfj** [comment]
> These kinds of utterancers are normally filtered out somehow

Yes, we have the following line in train.py

https://github.com/k2-fsa/icefall/blob/693f069de73fd91d7c2009571245d97221cc3a3f/egs/librispeech/ASR/zipformer/train.py#L1425

Not sure if @redsheep913  removes that or not.

**Comment by redsheep913** [comment]
@csukuangfj @danpovey I have not removed that line. It remains unchanged in train.py. 
I’m using a byte-level BPE model for the tokenizer; could this possibly be related?

**Comment by AmirHussein96** [comment]
@redsheep913 Looking at your data statistics, most recordings are quite short, though having a minimum around 0.5 seconds is still acceptable. You could try increasing the minimum duration threshold for filtering—for example, from 0.1 to 1.0 seconds:

`if c.duration <= 1.0 or c.duration >= 30.0`

This might help filter out short clips and give you a clearer picture. After that, you can focus on how to handle the shorter utterances.

**Comment by AmirHussein96** [comment]
@redsheep913 also what is your `params.output_downsampling_factor`? I think I missed to add this explicitly to the ASR since it was previously default to 2, please add the following to the ASR training command: `output-downsampling-factor=2`.

**Comment by redsheep913** [comment]
@AmirHussein96 

Thank you for your suggestions.

Previously, I used the following duration filter:
```python
if c.duration < 0.1 or c.duration > 30.0:
    logging.warning(
         f"Exclude cut with ID {c.id} from training. Duration: {c.duration}"
     )
    return False
```
I also implemented a check to filter out short token sequences:
```python
if T < len(tokens):
    logging.warning(
        f"Exclude cut with ID {c.id} from training. "
        f"Number of frames (before subsampling): {c.num_frames}. "
        f"Number of frames (after subsampling): {T}. "
        f"Text: {c.supervisions[0].text}. "
        f"Tokens: {tokens}. "
        f"Number of tokens: {len(tokens)}"
    )
    return False
```
However, I did not observe any warning messages being triggered from this check.

Additionally, I was able to complete RNN-T training and decoding without issues.

Regarding the `output-downsampling-factor` parameter, it is currently set to the default value of 2.

Please let me know if there is anything I might have overlooked, or if further adjustments would be recommended. Thank you.

**Comment by danpovey** [comment]
You could also print the lengths of utterances and of the embeddings after the frontend, to check whether a too-long utterance is causing a problem, there could be a mismatch between the filtering code and the actual execution code.

**Comment by KarelVesely84** [comment]
Actually, the minimum length of encoder output for CTC should be a little different.

There might be consecutive repetition of symbol in the reference,
and for this the CTC alignment must put a blank in between, so the 
reverse mapping of aligned symbols produces the original reference.

I realised it recently while playing with CTC aligner from torachaudio 
with the noisy yodas2 dataset.

I'll create a PR with a fix into the librispeech receipe.

Cheers!
Karel

**Comment by KarelVesely84** [comment]
see https://github.com/k2-fsa/icefall/pull/2051
